src/status_im/transport/message/core.cljs

```diff
@@ -64,36 +61,42 @@
       (aget obj i))
     [obj]))
 
-(fx/defn receive-whisper-messages
-  [{:keys [now] :as cofx} error messages chat-id]
-  (if (and (not error)
-           messages)
-    (let [now-in-s (quot now 1000)
-          receive-message-fxs (map (fn [message]
-                                     (receive-message now-in-s chat-id message))
-                                   messages)]
-      (apply fx/merge cofx receive-message-fxs))
-    (log/error "Something went wrong" error messages)))
-
-(fx/defn receive-messages [cofx event-js]
-  (let [fxs (keep
-             (fn [message-specs]
-               (let [chat (.-chat message-specs)
-                     messages (.-messages message-specs)
-                     error (.-error message-specs)]
-                 (when (seq messages)
-                   (receive-whisper-messages
-                    error
-                    messages
-                  ;; For discovery and negotiated filters we don't
-                  ;; set a chatID, and we use the signature of the message
-                  ;; to indicate which chat it is for
-                    (if (or (.-discovery chat)
-                            (.-negotiated chat))
-                      nil
-                      (.-chatId chat))))))
-             (.-messages event-js))]
-    (apply fx/merge cofx fxs)))
+(handlers/register-handler-fx
+ ::process
+ (fn [cofx [_ messages now-in-s]]
+   (let [[chat-id message] (first messages)
+         remaining-messages (rest messages)]
+     (if (seq remaining-messages)
+       (assoc
+        (receive-message cofx now-in-s chat-id message)
+        ;; We dispatch later to let the UI thread handle events, without this
+        ;; it will keep processing events ignoring user input.
+        :dispatch-later [{:ms 20 :dispatch [::process remaining-messages now-in-s]}])
+       (receive-message cofx now-in-s chat-id message)))))
+
+(fx/defn receive-messages
+  "Initialize the ::process event, which will process messages one by one
+  dispatching later to itself"
+  [{:keys [now] :as cofx} event-js]
+  (let [now-in-s (quot now 1000)
+        events (reduce
+                (fn [acc message-specs]
+                  (let [chat (.-chat message-specs)
+                        messages (.-messages message-specs)
+                        error (.-error message-specs)
+                        chat-id (if (or (.-discovery chat)
+                                        (.-negotiated chat))
+                                  nil
+                                  (.-chatId chat))]
+                    (if (seq messages)
+                      (reduce (fn [acc m]
+                                (conj acc [chat-id m]))
+                              acc
+                              messages)
+                      acc)))
+                []
+                (.-messages event-js))]
```

Comment:
Can we use beans here as well?

Reply:
beans is much slower then pure javascript from our tests, so I would avoid in critical paths , but I can double check if that's still the case

Reply:
ok, got it